experiments/run_single_experiment.py

```diff
@@ -25,56 +26,138 @@ def _main(cfg: DictConfig) -> None:
+        # Check if it's time to eval.
+        if t % cfg.eval_frequency == 0:
+            # Save the models from the training approach and load them into the
+            # eval approach.
+            step_model_dir = model_dir / str(t)
+            step_model_dir.mkdir(exist_ok=True)
+            train_approach.save(step_model_dir)
+            eval_approach.load(step_model_dir)
+            # Run evaluation.
+            step_eval_metrics = _evaluate_approach(eval_approach, eval_env, cfg, t)
+            eval_metrics.append(step_eval_metrics)
+        # Eval on the last time step but don't train anymore.
+        if t >= cfg.max_environment_steps:
+            break
+        # Continue training.
         try:
-            act = approach.step()
+            act = train_approach.step()
         except ApproachFailure as e:
             logging.info(e)
-        obs, rew, terminated, truncated, info = env.step(act)
+        obs, rew, terminated, truncated, info = train_env.step(act)
         assert np.isclose(rew, 0.0)
         assert not (terminated or truncated)
-        approach.update(obs, float(rew), terminated, info)
+        train_approach.update(obs, float(rew), terminated, info)
         user_satisfaction = info.get("user_satisfaction", np.nan)
-        env_video_should_pause = info.get("env_video_should_pause", False)
-        step_metrics = {
+        step_train_metrics = {
             "step": t,
-            "user_allows_explore": info["user_allows_explore"],
             "user_satisfaction": user_satisfaction,
-            "env_video_should_pause": env_video_should_pause,
-            **approach.get_step_metrics(),
+            **train_approach.get_step_metrics(),
         }
         logging.info(f"Step {t} satisfaction: {user_satisfaction}")
-        metrics.append(step_metrics)
-    env.close()
+        train_metrics.append(step_train_metrics)
+    train_env.close()
+    eval_env.close()
 
     # Aggregate and save results.
-    df = pd.DataFrame(metrics)
-    df.to_csv(cfg.results_file)
-    logging.info(f"Wrote out results to {cfg.results_file}")
+    train_df = pd.DataFrame(train_metrics)
+    train_df.to_csv(cfg.train_results_file)
+    logging.info(f"Wrote out training results to {cfg.train_results_file}")
+
+    eval_df = pd.DataFrame(eval_metrics)
+    eval_df.to_csv(cfg.eval_results_file)
+    logging.info(f"Wrote out eval results to {cfg.eval_results_file}")
+
+
+def _evaluate_approach(
+    eval_approach: BaseApproach, eval_env: gym.Env, cfg: DictConfig, training_step: int
+) -> dict[str, float]:
+    """Evaluate the given approach and return metrics."""
+    # Evaluate for a given number of trials.
+    cumulative_user_satisfactions: list[float] = []
+    for eval_trial_idx in range(cfg.num_eval_trials):
+        logging.info(f"Starting eval trial {eval_trial_idx}")
+        seed = cfg.seed + cfg.eval_seed_offset + eval_trial_idx
+        obs, info = eval_env.reset(seed=seed)
```

Comment:
Just making sure, does the code ever rely on sampling directly from `env.action_space`? I didn't see any related usage, but it seems that in the experiment setup, both `train_env.action_space` and `eval_env.action_space` were explicitly seeded, but not here.